src/expr/src/relation/mod.rs

```diff
@@ -338,23 +339,11 @@ impl RelationExpr {
                     .collect::<Vec<_>>();
                 let mut typ = RelationType::new(column_types);
 
-                let input_arities = input_types
-                    .iter()
-                    .map(|i| i.column_types.len())
-                    .collect::<Vec<_>>();
-
-                let mut offset = 0;
-                let mut prior_arities = Vec::new();
-                for input in 0..inputs.len() {
-                    prior_arities.push(offset);
-                    offset += input_arities[input];
-                }
-
-                let input_relation = input_arities
-                    .iter()
-                    .enumerate()
-                    .flat_map(|(r, a)| std::iter::repeat(r).take(*a))
-                    .collect::<Vec<_>>();
+                // It is important the `new_from_input_types` constructor is
+                // used. Otherwise, Materialize may potentially end up in an
+                // infinite loop.
+                let input_mapper =
+                    join_input_mapper::JoinInputMapper::new_from_input_types(&input_types);
 
                 // A relation's uniqueness constraint holds if there is a
```

Comment:
By the way, does this hold anymore given that we no longer reorder join inputs after deciding on an implementation?

Reply:
I'm not sure I understand! I think the statement of the comment is true about a join, independent of the implementation or the order of the inputs. It just means to say, if you have A x B x C then if B and C have unique keys that are used appropriately in the join, then the join result has the same keys as in A. I think this reasoning is true starting from A, or B, or C.

Reply:
I see. I was confused by why it is always the case that the only key potentially retained is from the first input. I guess it's that the uniqueness constraint holds regardless of join input order, but we currently choose to only detect retained keys if it came from the first input.

So if we had:
`create table foo (a primary key, b)`
`create table bar (a primary key, b)`
and a RelationExpr::Join with the inputs in the order `[foo, bar]`:
If the join constraint is `foo.b = bar.a`, then we know that `foo.a` remains unique.
If the join constraint is `foo.a = bar.b`, then `bar.a` will remain unique, but this method does not detect that.

Reply:
That's right!